lib/util/addDevServerEntrypoints.js

```diff
@@ -19,16 +19,22 @@ module.exports = function addDevServerEntrypoints(webpackOptions, devServerOptio
 
     if (devServerOptions.hotOnly) { devClient.push('webpack/hot/only-dev-server'); } else if (devServerOptions.hot) { devClient.push('webpack/hot/dev-server'); }
 
-    [].concat(webpackOptions).forEach((wpOpt) => {
-      if (typeof wpOpt.entry === 'object' && !Array.isArray(wpOpt.entry)) {
-        Object.keys(wpOpt.entry).forEach((key) => {
-          wpOpt.entry[key] = devClient.concat(wpOpt.entry[key]);
+    const prependDevClient = (entry) => {
+      if (typeof entry === 'function') {
+        return () => Promise.resolve(entry()).then(prependDevClient);
```

Comment:
This will handle both functions which return entry objects and functions which return a `Promise` due to the behavior of `Promise.resolve`.  It is the same technique used in the `DynamicEntryPlugin` itself ([link](https://github.com/webpack/webpack/blob/213226ece261a0e89edb9de15fbc6775418f6041/lib/DynamicEntryPlugin.js#L49)).